app/uk/gov/hmrc/tradergoodsprofiles/controllers/actions/ValidateHeaderAction.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tradergoodsprofiles.controllers.actions
+
+import cats.data.EitherT
+import play.api.http.{HeaderNames, MimeTypes}
+import play.api.mvc.{ActionFilter, Request, Result}
+import uk.gov.hmrc.tradergoodsprofiles.config.Constants
+import uk.gov.hmrc.tradergoodsprofiles.models.errors.InvalidHeaderErrorResponse
+import uk.gov.hmrc.tradergoodsprofiles.services.DateTimeService
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class ValidateHeaderAction @Inject() (datetimeService: DateTimeService)(implicit ec: ExecutionContext)
+    extends ActionFilter[Request] {
+
+  override val executionContext: ExecutionContext = ec
+
+  override def filter[A](request: Request[A]): Future[Option[Result]] = {
+
+    val result = for {
+      _ <- validateAcceptHeader(request)
+      _ <- validateContentTypeHeader(request)
+      _ <- validateClientIdHeader(request)
+    } yield None
+
+    result.merge
+  }
+
+  private def validateAcceptHeader(request: Request[_]): EitherT[Future, Option[Result], Unit] = {
+    val pattern = """^application/vnd[.]{1}hmrc[.]{1}1{1}[.]0[+]{1}json$""".r
```

Comment:
we can move this outside of the function and set it to a lazy val so it isn't generated every time the function is called

Why is the left an Option?

Reply:
Yes I see your point, but also there is the argument there which is keep the code that work together stay together, but I am not particularly fussy in this case.

That is an option because it derived from ActionFilter and override the filter function which return an Option[Result]

Reply:
But the option here will always be set, so what's the point in returning an option when it's always a Some

Reply:
is not always a some. If the validation is fine, it will return a None which means carry on. Action filter will intercept the request and if the validation is fine it will continue process (return None), otherwise return an error

Reply:
Got it